stumpy/fastmath.py

```diff
@@ -0,0 +1,89 @@
+import ast
+import importlib
+import pathlib
+
+from stumpy import config
+
+
+def get_default_fastmath(module_name):
```

Comment:
There are many nested if & for. I was wondering if we can revise this function as follows:

```
def get_default_fastmath(module_name, func_name):
    filepath = pathlib.Path(__file__).parent / f"{module_name}.py"
    file_contents = ""
    with open(filepath, encoding="utf8") as f:
        file_contents = f.read()

    func_node = None
    module = ast.parse(file_contents)
    for node in module.body:
        if isinstance(node, ast.FunctionDef) and node.name == func_name:
            func_node = node
            break

    if func_node is None:
        msg = f"The module `{module_name}` does not have function `{func_name}`"
        raise ValueError(msg)

    decorator_node = None
    for decorator in func_node.decorator_list:
        if (
            isinstance(decorator, ast.Call)
            and isinstance(decorator.func, ast.Name)
            and decorator.func.id == "njit"
        ):
            decorator_node = decorator
            break

    if decorator_node is None:
        msg = (
            f"The function `{func_name}` in module `{module_name}`"
            + "does not have njit decorator"
        )
        raise ValueError(msg)

    fastmath_default = None
    for keyword in decorator_node.keywords:
        if keyword.arg == "fastmath":
            config_var = keyword.value.attr
            fastmath_default = config._STUMPY_DEFAULTS[config_var]
            break

    return fastmath_default
```

Reply:
Maybe we should revise the function `cache.get_njit_funcs` as well to reduce the number of nested if-block & for loops. Also, that function and the one provided in my previous comment seem to have some common parts. So, curious to see if we can do some refactoring.

Reply:
Wait, why are you doing it this way? This seems overly complicated. I think we can actually add the defaults to `config.py`:

```
# config.py

import numpy as np
import warnings
import . import cache
import importlib

_STUMPY_DEFAULTS = {
    "STUMPY_THREADS_PER_BLOCK": 512,
    "STUMPY_MEAN_STD_NUM_CHUNKS": 1,
    "STUMPY_MEAN_STD_MAX_ITER": 10,
    "STUMPY_DENOM_THRESHOLD": 1e-14,
    "STUMPY_STDDEV_THRESHOLD": 1e-7,
    "STUMPY_P_NORM_THRESHOLD": 1e-14,
    "STUMPY_TEST_PRECISION": 5,
    "STUMPY_MAX_P_NORM_DISTANCE": np.finfo(np.float64).max,
    "STUMPY_MAX_DISTANCE": np.sqrt(np.finfo(np.float64).max),
    "STUMPY_EXCL_ZONE_DENOM": 4,
    "STUMPY_FASTMATH": True,
    "STUMPY_FASTMATH_FLAGS": {"nnan", "ninf", "nsz", "arcp", "contract", "afn", "reassoc"},
}

STUMPY_THREADS_PER_BLOCK = _STUMPY_DEFAULTS["STUMPY_THREADS_PER_BLOCK"]
STUMPY_MEAN_STD_NUM_CHUNKS = _STUMPY_DEFAULTS["STUMPY_MEAN_STD_NUM_CHUNKS"]
STUMPY_MEAN_STD_MAX_ITER = _STUMPY_DEFAULTS["STUMPY_MEAN_STD_MAX_ITER"]
STUMPY_DENOM_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_DENOM_THRESHOLD"]
STUMPY_STDDEV_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_STDDEV_THRESHOLD"]
STUMPY_P_NORM_THRESHOLD = _STUMPY_DEFAULTS["STUMPY_P_NORM_THRESHOLD"]
STUMPY_TEST_PRECISION = _STUMPY_DEFAULTS["STUMPY_TEST_PRECISION"]
STUMPY_MAX_P_NORM_DISTANCE = _STUMPY_DEFAULTS["STUMPY_MAX_P_NORM_DISTANCE"]
STUMPY_MAX_DISTANCE = _STUMPY_DEFAULTS["STUMPY_MAX_DISTANCE"]
STUMPY_EXCL_ZONE_DENOM = _STUMPY_DEFAULTS["STUMPY_EXCL_ZONE_DENOM"]
STUMPY_FASTMATH = _STUMPY_DEFAULTS["STUMPY_FASTMATH"]
STUMPY_FASTMATH_FLAGS = _STUMPY_DEFAULTS["STUMPY_FASTMATH_FLAGS"]

# Add fastmath defaults for each njit function
njit_funcs = cache.get_njit_funcs()
for module_name, func_name in njit_funcs:
    module = importlib.import_module(f".{module_name}", package="stumpy")
    func = getattr(module, func_name)
    key = module_name + '.' + func_name
    key = 'STUMPY_' + key.upper()
    _STUMPY_DEFAULTS[key] = func.targetoptions['fastmath']
```

This approach automatically adds the `fastmath` flags for each njitted function directly to `_STUMPY_DEFAULTS`, which is a natural/obvious place to put the default values.

Reply:
This is indeed less complicated! I do not know how the last line, i.e. `_STUMPY_DEFAULTS[key] = func.targetoptions['fastmath']` behaves if a func is recompiled with new fastmath flags and then an `import config` occurs. I am curious to check that.

Currently, I am trying to resolve an issue raised by circular imports.

Reply:
> Currently, I am trying to resolve an issue raised by circular imports.

If this is the case then you may need to move something into `core.py` instead.

Reply:
> I do not know how the last line, i.e. _STUMPY_DEFAULTS[key] = func.targetoptions['fastmath'] behaves if a func is recompiled with new fastmath flags and then an import config occurs. I am curious to check that.

If you are concerned that changing `func.targetoptions['fastmath']` to some other value will also change `_STUMPY_DEFAULTS[key]` then it appears that it won't based on some simple checks that I did. But please verify for yourself.

Reply:
> If this is the case then you may need to move something into core.py instead.

Thanks for mentioning that. I remember we did something similar in [another case](https://github.com/TDAmeritrade/stumpy/pull/714/commits/d4dc9200aa37ca05be44583de7288a4665bbe567). Going to try and report back.